### PR TITLE
[CI] Golang Precommit hooks

### PR DESCRIPTION
**Pull request by segunjkf**
**Notes for Reviewers**

This PR fixes #

**[Signed commits](https://github.com/meshery/meshery/blob/master/CONTRIBUTING.md#signing-off-on-commits-developer-certificate-of-origin)**
- [x] Yes, I signed my commits.

### CODE REVIEW

**Comment by Chadha93** [comment]
@segunjkf  Let's discusss this in tomorrow's [Meshery Dev call](https://layer5.io/community/calendar). Please add this as an agenda item, in the Meshery [meeting minutes](https://docs.google.com/document/d/1Dg4Uz9rGTuGHmYZwIt47z1j5x3LmnpxD1M1SgypAkQ0/edit#) for feedback/review.

**Comment by acald-creator** [comment]
+1

**Comment by segunjkf** [comment]
hello seems there no activity here? @leecalcote

**Comment by acald-creator** [comment]
@segunjkf Was this discussed at all on the Meshery dev call from last mentioned?

**Comment by leecalcote** [comment]
@theBeginner86, we're overdue on providing @segunjkf some meaningful feedback. Given your pre-commit hook work in Meshery Cloud, I wonder if you might review this PR and see about accepting @segunjkf's work. 👍

**Comment by leecalcote** [comment]
@segunjkf in relation to the items of consideration linked to here - https://github.com/meshery/meshery/issues/6568#issuecomment-1327782151, I wonder what thoughts you have on considerations we should be making.

File: .dccache [hunk 1]
```diff
@@ -0,0 +1 @@
...patibility/meshery-linkerd/03-22-2022-20-28-03_.html":[2,1668015253958.047,"7eb70257593da06f682a3ddda54a9d260d4fc514f645237f5ca74b08f8da61a6"],"c:/Users/SEGUN/Desktop/Contribution/meshery/docs/v0.5/project/compatibility/meshery-linkerd/03-25-2022-23-17-03_.html":[2,1668015253960.0518,"7eb70257593da06f682a3ddda54a9d260d4fc514f645237f5ca74b08f8da61a6"],"c:/Users/SEGUN/Desktop/Contribution/meshery/docs/v0.5/project/compatibility/meshery-linkerd/03-29-2022-16-54-03_.html":[2,1668015253962.0562,"7eb70257593da06f682a3ddda54a9d260d4fc514f645237f5ca74b08f8da61a6"],"c:/Users/SEGUN/Desktop/Contribution/meshery/docs/v0.5/project/compatibility/meshery-linkerd/03-30-2022-21-37-03_.html":[2,1668015253965.0515,"7eb70257593da06f682a3ddda54a9d260d4fc514f645237f5ca74b08f8da61a6"],"c:/Users/SEGUN/Desktop/Contribution/meshery/docs/v0.5/project/compatibility/meshery-linkerd/04-07-2022-13-29-04_.html":[2,1668015253967.0493,"7eb70257593da06f682a3ddda54a9d260d4fc514f645237f5ca74b08f8da61a6"],"c:/Users/SEGUN/Desktop/Contribution/meshery/docs/v0.5/project/compatibility/meshery-linkerd/04-15-2022-22-28-04_.html":[2,1668015253969.0505,"7eb70257593da06f682a3ddda54a9d260d4fc514f645237f5ca74b08f8da61a6"],"c:/Use
```
Comment (theBeginner86) [inline .dccache:1]:
```suggestion

```

Comment (theBeginner86) [reply .dccache:1]: We should probably be adding this to `.gitignore`?

**Review by theBeginner86 — COMMENTED** [review]
@segunjkf 
Thanks for this PR.

I have a couple of questions/considerations:

1. We need to update the documentation to ensure that contributors install the [pre-commit](https://pre-commit.com/#install) CLI as this doesn't work out of the box.
2. We need to update documentation regarding the following error from the pre-commit hook:
```[ERROR] Cowardly refusing to install hooks with core.hooksPath set.``` 
I got this error too. The reason is that if git is already tracking a pre-commit hook like `ui/.husky` in my case (most of the other UI contributors) then they need to unset it from their git config (set it back if they want to contribute to UI).
Reference: https://stackoverflow.com/questions/67793193/cowardly-refusing-to-install-hooks-with-core-hookspath-set
We possibly need to update our hook strategy or deprecate the use of husky, and just use a UI plugin from pre-commit itself, like https://github.com/pre-commit/mirrors-eslint.
There seem some added configurations that need to be done as I wasn't able to use that plugin instantly for UI. (was conflicting with our Babel config)
I believe a an `Make` command to switch between hookPaths would be much better and we can keep using the husky hook which we have established in this and other repo.
3. This PR should also address and fix all the backend **lint checks + failing unit tests** because if we merge this then new backend contributors wouldn't be able to add new commits until they fix those checks. eg: _A couple of `mesheryctl` unit tests are failing._
4. We can get away with most of the configuration stuff by updating/creating new `Make` commands. 

Thoughts on this @segunjkf @leecalcote?

**Comment by Chadha93** [comment]
@segunjkf  Let's discuss this on the Meshery Dev call. Please add this as an agenda item in the [meeting minutes](https://docs.google.com/document/d/1Dg4Uz9rGTuGHmYZwIt47z1j5x3LmnpxD1M1SgypAkQ0/edit#) if you would. :)

**Comment by leecalcote** [comment]
> @segunjkf Let's discuss this on the Meshery Dev call. Please add this as an agenda item in the [meeting minutes](https://docs.google.com/document/d/1Dg4Uz9rGTuGHmYZwIt47z1j5x3LmnpxD1M1SgypAkQ0/edit#) if you would. :)

It would be helpful to get a few more eyes on this.

**Comment by leecalcote** [comment]
How do we ensure that this precommit hook only runs on change to .go files?

**Comment by segunjkf** [comment]
> > @segunjkf Let's discuss this on the Meshery Dev call. Please add this as an agenda item in the [meeting minutes](https://docs.google.com/document/d/1Dg4Uz9rGTuGHmYZwIt47z1j5x3LmnpxD1M1SgypAkQ0/edit#) if you would. :)
> 
> It would be helpful to get a few more eyes on this.

yes sure will do that.

**Comment by segunjkf** [comment]
> @segunjkf Thanks for this PR.
> 
> I have a couple of questions/considerations:
> 
> 1. We need to update the documentation to ensure that contributors install the [pre-commit](https://pre-commit.com/#install) CLI as this doesn't work out of the box.
> 2. We need to update documentation regarding the following error from the pre-commit hook:
>    `[ERROR] Cowardly refusing to install hooks with core.hooksPath set.`
>    I got this error too. The reason is that if git is already tracking a pre-commit hook like `ui/.husky` in my case (most of the other UI contributors) then they need to unset it from their git config (set it back if they want to contribute to UI).
>    Reference: https://stackoverflow.com/questions/67793193/cowardly-refusing-to-install-hooks-with-core-hookspath-set
>    We possibly need to update our hook strategy or deprecate the use of husky, and just use a UI plugin from pre-commit itself, like https://github.com/pre-commit/mirrors-eslint.
>    There seem some added configurations that need to be done as I wasn't able to use that plugin instantly for UI. (was conflicting with our Babel config)
>    I believe a an `Make` command to switch between hookPaths would be much better and we can keep using the husky hook which we have established in this and other repo.
> 3. This PR should also address and fix all the backend **lint checks + failing unit tests** because if we merge this then new backend contributors wouldn't be able to add new commits until they fix those checks. eg: _A couple of `mesheryctl` unit tests are failing._
> 4. We can get away with most of the configuration stuff by updating/creating new `Make` commands.
> 
> Thoughts on this @segunjkf @leecalcote?

we should discuss more on the meeting

**Comment by segunjkf** [comment]
> How do we ensure that this precommit hook only runs on change to .go files?



> How do we ensure that this pre-commit hook only runs on change to .go files?

The Go pre-commit hooks repo used by default it is set to run on any file with .go extension `https://github.com/dnephin/pre-commit-golang/blob/master/.pre-commit-hooks.yaml` but it also possible to specify `files: "^.*\\.go$"` to ensure it only runs on files with go extension. but by default it runs on only files with go extension

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -148,6 +148,15 @@ To test code changes on your local machine, run the following command:
 make golangci-run
 ```
 
+#### Golang Pre-commit
+Before committing their code, users can now see the potential errors. To use this feature, you must first install Pre-commit. Please follow these [download instructions](https://pre-commit.com/).
+
+Please run the following commands to install the pre-commit hooks on your local machine.
```
Comment (theBeginner86) [inline CONTRIBUTING.md:154]:
```suggestion
Once installed, please run the following command to set up the configured git hook scripts. 
```

**Review by theBeginner86 — APPROVED** [review]

**Comment by theBeginner86** [comment]
LGTM

Thanks @segunjkf for the docs update

**Comment by segunjkf** [comment]
@leecalcote i think this pull request is ready to be merged.

**Comment by acald-creator** [comment]
Before we merge this, I just checked out the repo at `github.com/dnephin/pre-commit-golang`.

Please take a look at this issue: https://github.com/dnephin/pre-commit-golang/issues/98

The person over this project is going to stop maintaining this project. I believe there are better alternatives, and I can help with the research in the meantime.

**Comment by acald-creator** [comment]
I would highly suggest looking at this project: `https://github.com/TekWizely/pre-commit-golang`?

**Comment by segunjkf** [comment]
> Before we merge this, I just checked out the repo at `github.com/dnephin/pre-commit-golang`.
> 
> Please take a look at this issue: [dnephin/pre-commit-golang#98](https://github.com/dnephin/pre-commit-golang/issues/98)
> 
> The person over this project is going to stop maintaining this project. I believe there are better alternatives, and I can help with the research in the meantime.
> 
> The good thing is that since we're already using `golangci-lint`, I found a couple examples where you can use `golangci-lint` as pre-commit. I'll see if I can make some changes on this PR.

I will also look into this. Thank you

**Comment by acald-creator** [comment]
@segunjkf Do you have any updates to provide for this PR?

**Comment by money8203** [comment]
Is this pr still in progress ?

**Comment by segunjkf** [comment]
> @segunjkf Do you have any updates to provide for this PR?

sorry I saw this late. I look into it and provide feedback.

**Comment by Chadha93** [comment]
@segunjkf  Let's discuss this on the Meshery Dev call. Please add this as an agenda item in the [meeting minutes](https://docs.google.com/document/d/1Dg4Uz9rGTuGHmYZwIt47z1j5x3LmnpxD1M1SgypAkQ0/edit#) if you would. :)

**Comment by Azanul** [comment]
@segunjkf [This](https://github.com/golangci/golangci-lint/blob/master/.pre-commit-hooks.yaml) is how golangci-lint does it.

**Pull request closed without merging**